appsTransfer/src/main/java/org/kinecosystem/appstransfer/view/customview/AppsTransferList.kt

```diff
@@ -0,0 +1,167 @@
+    }
+
+    override fun navigateToUrl(googlePlayUrl: String) {
+        context.navigateToUrl(googlePlayUrl)
+    }
+
+    override fun updateData(apps: List<EcosystemApp>) {
+        (adapter as AppsTransferAdapter).updateData(apps)
+    }
+
+    override fun setLoadingListener(loadingListener: AppsTransferListPresenter.LoadingListener) {
+        presenter?.setLoadingListener(loadingListener)
+    }
+
+    override fun transferToApp(app: EcosystemApp) {
+        //TODO
+        Log.d("###", "### start transfer to app ${app.identifier}")
+        // context.startActivity(AppInfoActivity.getIntent(context, app.name))
+    }
+
+    override fun onWindowFocusChanged(hasWindowFocus: Boolean) {
+        super.onWindowFocusChanged(hasWindowFocus)
+        if (hasWindowFocus) {
+            Log.d("###", "### list on resumed")
+            (adapter as AppsTransferAdapter).invalidateData()
+            Log.d("###", "###  after notifyDataSetChanged")
+        } else {
+            Log.d("###", "### list on puaused")
+
+        }
+    }
+
+    override fun onAttachedToWindow() {
+        super.onAttachedToWindow()
+        presenter?.onAttach(this)
+    }
+
+    override fun onDetachedFromWindow() {
+        super.onDetachedFromWindow()
+        presenter?.onDetach()
+    }
+}
+
+class AppsTransferAdapter(private val context: Context, private val presenter: IAppsTransferListPresenter) : RecyclerView.Adapter<AppsTransferAdapter.ViewHolder>() {
+
+    var apps: List<EcosystemApp> = listOf()
+
+    fun updateData(apps: List<EcosystemApp>) {
+        this.apps = apps.sortedByDescending { context.isAppInstalled(it.identifier!!) }
+        notifyDataSetChanged()
+    }
+
+    fun invalidateData(){
+        //TODO check sort
+        apps = apps.sortedByDescending { context.isAppInstalled(it.identifier!!) }
+        notifyDataSetChanged()
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(parent?.context)
+                .inflate(R.layout.app_transfer_item, parent, false)
+        return ViewHolder(context, view)
+    }
+
+    override fun getItemCount(): Int {
+        return apps.size
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
+        val app: EcosystemApp = apps[position]
+        holder?.bind(app)
+        holder?.actionText?.setOnClickListener {
+            presenter.onAppClicked(app)
+        }
+    }
+
+    class ViewHolder(private val context: Context, val view: View) : RecyclerView.ViewHolder(view) {
+        private val icon: ImageView = view.findViewById(R.id.appIcon)
+        private val appName: TextView = view.findViewById(R.id.appName)
+        private val status: TextView = view.findViewById(R.id.status)
+        val actionText: TextView = view.findViewById(R.id.actionText)
+
+        fun bind(app: EcosystemApp) {
+            appName.text = app.name
+            icon.load(app.iconUrl)
+            app.identifier?.let {
```

Comment:
In `updateData` and `invalidateData` methods you use it.identifier!! 
Do you need the let here?

Reply:
yes because of the app